### ISSUE TITLE
crash on resizing Linux Nvidia

### ISSUE THREAD

**Issue by danilw** [KhronosGroup/Vulkan-Samples#250]
## this is not a Nvidia bug, read https://github.com/KhronosGroup/Vulkan-Samples/issues/250#issuecomment-1416238684

Vulkan SDK 1.2.131.2
Vulkan-Samples cloned and built just now
Nvidia driver latest 460.39 Linux

**Bug:**
Crash on resizing (after multiple resizes, many times resize window)

launching and resizing window
`build/linux/app/bin/Release/x86_64/vulkan_samples --sample texture_loading`
result 
```
[error] [samples/api/texture_loading/texture_loading.cpp:465] Detected Vulkan error: ERROR_DEVICE_LOST
```

launching with `VK_INSTANCE_LAYERS=VK_LAYER_KHRONOS_validation`
*vk_layer_settings.txt* config:
```
khronos_validation.debug_action = VK_DBG_LAYER_ACTION_LOG_MSG
khronos_validation.report_flags = error,warn,perf
khronos_validation.log_filename = log.txt
```

full log [log.txt](https://github.com/KhronosGroup/Vulkan-Samples/files/6018152/log.txt)

Mip errors
```VUID-VkFramebufferCreateInfo-pAttachments-00882(ERROR / SPEC): msgNum: 0 - vkCreateFramebuffer(): VkFramebufferCreateInfo attachment #0 mip level 0 has dimensions smaller than the corresponding framebuffer dimensions. Here are the respective dimensions for attachment #0, framebuffer:```
come without crashing just on resizing

crash is:
```
VUID-vkDestroyImageView-imageView-01026(ERROR / SPEC): msgNum: 0 - Cannot call vkDestroyImageView on VkImageView 0xaec0000000aec[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to imageView must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyImageView-imageView-01026)
    Objects: 1
        [0] 0xaec0000000aec, type: 14, name: NULL
VUID-vkDestroyImageView-imageView-01026(ERROR / SPEC): msgNum: 0 - Cannot call vkDestroyImageView on VkImageView 0xaf10000000af1[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to imageView must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyImageView-imageView-01026)
    Objects: 1
        [0] 0xaf10000000af1, type: 14, name: NULL
VUID-vkDestroyImage-image-01000(ERROR / SPEC): msgNum: 0 - Cannot call vkDestroyImage on VkImage 0xaef0000000aef[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to image, either directly or via a VkImageView, must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyImage-image-01000)
    Objects: 1
        [0] 0xaef0000000aef, type: 10, name: NULL
VUID-vkFreeMemory-memory-00677(ERROR / SPEC): msgNum: 0 - Cannot call vkFreeMemory on VkDeviceMemory 0xaf00000000af0[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to memory (via images or buffers) must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeMemory-memory-00677)
    Objects: 1
        [0] 0xaf00000000af0, type: 8, name: NULL
VUID-vkDestroyFramebuffer-framebuffer-00892(ERROR / SPEC): msgNum: 0 - Cannot call vkDestroyFramebuffer on VkFramebuffer 0xaf20000000af2[] that is currently in use by a command buffer. The Vulkan spec states: All submitted commands that refer to framebuffer must have completed execution (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkDestroyFramebuffer-framebuffer-00892)
    Objects: 1
        [0] 0xaf20000000af2, type: 24, name: NULL
VUID-vkFreeCommandBuffers-pCommandBuffers-00047(ERROR / SPEC): msgNum: 0 - Attempt to free VkCommandBuffer 0x2619cd8[] which is in use. The Vulkan spec states: All elements of pCommandBuffers must not be in the pending state (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkFreeCommandBuffers-pCommandBuffers-00047)
    Objects: 1
        [0] 0x2619cd8, type: 6, name: NULL
```

**Comment by pdaniell-nv** [comment]
I wasn't able to reproduce this on Windows. Are you able to reproduce it with the latest beta drivers Windows 457.96 or Linux 455.50.07 available from here: https://developer.nvidia.com/vulkan-driver.

**Comment by danilw** [comment]
**this is only Linux problem**, Windows use other VK_surface

**this bug works only on `vkCreateXcbSurfaceKHR`** this XCB surface (X11 only) (maybe its only X11 bug)

Linux:
OS - OpenSUSE Tumbleweed
kernel 5.10.1-1
xorg-x11-server 1.20.10-1.1
DE - tested on XFCE4 and Gnome3 (crash same, DE does not matter I think)

reason (I think, can be wrong) - resize event to application come with wrong values(like outdated resize event when many resize events happens(when user keep resizing window)) 

**I trigger this crash after 20-50 sec of resizing window it always crashes** (I mean it random, can be even first 2 sec of resizing crash)

> Are you able to reproduce it with the latest beta drivers Windows 457.96 or Linux 455.50.07

**I have test and confirm** this bug on:
*Latest Production Branch Version: 460.56* just installed (look screenshot)
![1](https://user-images.githubusercontent.com/24825887/109575635-4b194e00-7b03-11eb-989a-3b0fe3a24ea4.png)

that *Linux 455.50.07* driver does not work for me for some reason, it compiled installed and X11 can not start on this driver

**does this bug/crash happens on other Vulkan applications**:
yes

other applications (Vulkan examples from VulkanSDK 1.2 and other github projects) report same error:

```
VUID-VkSwapchainCreateInfoKHR-imageExtent-01274(ERROR / SPEC): msgNum: 0 - vkCreateSwapchainKHR() called with imageExtent = (674,743), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (675,743), minImageExtent = (675,743), maxImageExtent = (675,743). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
```
but not every application crash

**Comment by pdaniell-nv** [comment]
Thanks for the extra details. I will pass this on our Linux engineers to investigate (nvbug 3268562).

**Comment by damienleone** [comment]
I looked into this issue from the NVIDIA driver perspective and to me it seems related to how Vulkan-Samples manages window resizes.

It appears that Vulkan-Samples monitors window resizes itself each frame per the following trace:
`prepare_frame -> handle_surface_changes -> vkGetPhysicalDeviceSurfaceCapabilitiesKHR`

It can therefore recreate a swapchain before receiving a VK_ERROR_OUT_OF_DATE_KHR from the driver.

It seems to me that there is a mismatch between what Vulkan-Samples' Swapchain class thinks the extent of the surface is vs. the ApiVulkanSample class.

I should note that I can only reproduce this when reducing the size of the window, but not when I increase it. My test case is to slowly increase/decrease the width of the application without touching its height.

I added a bunch of printfs in the app and the driver and noticed the below:
```
ApiVulkanSample::resize(): width = 1763 height = 720
handle_surface_changes: vkGetPhysicalDeviceSurfaceCapabilitiesKHR: 1763x720
handle_surface_changes: vkGetPhysicalDeviceSurfaceCapabilitiesKHR: 1763x720
handle_surface_changes: vkGetPhysicalDeviceSurfaceCapabilitiesKHR: 1763x720
handle_surface_changes: vkGetPhysicalDeviceSurfaceCapabilitiesKHR: 1752x720
handle_surface_changes: vkGetPhysicalDeviceSurfaceCapabilitiesKHR: 1752x720
handle_surface_changes: vkGetPhysicalDeviceSurfaceCapabilitiesKHR: 1736x720
[...]
Swapchain request_extent: width = 1736 height = 720
[...]
*** Creating swapchain 0x5559c98cffa0 1736x720, 3 images (oldSwapchain = 0x5559c7511ad0)
Vulkan drawable became out of date: 0x11
*** Destroying swapchain 0x5559c7511ad0
[info] Recreated swapchain
[...]
ApiVulkanSample::resize(): width = 1752 height = 720
[error] [framework/api_vulkan_sample.cpp:530] Detected Vulkan error: ERROR_DEVICE_LOST
```
We can see a mismatch in the bookkeeping, ApiVulkanSample thinks the surface is 1752x720 for a swapchain that has a surface extent of 1736x720.

I suspect this causes the app to render out of bounds, causing the DEVICE_LOST.

**Comment by danilw** [comment]
> It appears that Vulkan-Samples monitors window resizes itself each frame per the following trace:

this is not "only" Vulkan-Samples problem, almost every other Vulkan application handle resize event same as Khronos Vulkan Samples, and many crashes come because of this bug (as I saw)

I have no idea how to fix it in Vulkan-Samples (correctly), in my code I just add "dirt bad fix" just adding long Sleep time before actual resize.

Thanks for info, nice to know it.

**Comment by damienleone** [comment]
> this is not "only" Vulkan-Samples problem, almost every other Vulkan application handle resize event same as Khronos Vulkan Samples, and many crashes come because of this bug (as I saw)

From taking a quick look at vkcube, demo->{width,height} is the only source of truth for tracking the size of the application. It is updated on XCB_CONFIGURE_NOTIFY events and a swapchain is recreated with the size that was set then.

**Comment by cubanismo** [comment]
Also, note you can never actually reliably know the width/height of a window. It can change asynchronously in the window server at any time on most windowing systems. What a Vulkan application can rely on is that once it has acquired a swapchain image, that swapchain image is (1) the same size as the swapchain used to create it and (2) will remain valid until presented or the swapchain is destroyed. However, any other time, the application must be able to handle OUT_OF_DATE errors from functions acquiring or presenting swapchain images and recreate the swapchain and any other objects referencing swapchain images, potentially multiple times until the window server and application state are in sync long enough to acquire another swapchain image (I.e., the app should loop until it reaches a temporary steady state). On some implementations on some window systems, the act of creating a swapchain itself may cause the implementation to update its internal view of the surface and hence cause the swapchain returned to immediately be out of date. This may seem counterintuitive from an application point of view, but given the above-mentioned necessity for looping until a steady state is reached, there isn't actually any additional code required in the application to handle this case anyway.

> this is not "only" Vulkan-Samples problem, almost every other Vulkan application handle resize event same as Khronos Vulkan Samples, and many crashes come because of this bug (as I saw)

All the more reason to get the sample right and hopefully avoid this in future apps based on the sample.

**Comment by TomAtkinsonArm** [comment]
Can this be closed?

**Comment by TomAtkinsonArm** [comment]
Re-open if this issue persists

**Comment by tgfrerer** [comment]
I can confirm, unfortunately, that the issue seems to persist on Linux at least... 

I have just been able to reproduce this with NVIDIA Driver Version: `525.78.01`. and VulkanSamples git commit: 7e26046720669d82d19d4525f2bb7308efd4b55b

I start the app normally, then erratically resize the window until `VK_ERROR_DEVICE_LOST`, which happens after an unpredictable but short enough (a few seconds) time.

I'm attaching a log: [log.txt](https://github.com/KhronosGroup/Vulkan-Samples/files/10568162/log.txt)

What brought me to this issue is that I have seen what looks to be the same issue coming up when developing my renderer, [Island](https://github.com/tgfrerer/island/tree/wip_swapchain_dynamic). 

Placing a `vkQueueWaitIdle()` before the (re-creating) call to `vkCreateSwapchainKHR()` there seems to solve the issue, which might suggest that `vkCreateSwapchainKHR` touches some objects which are in-flight, for which I could not find any explanation reading the [spec](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkSwapchainCreateInfoKHR.html). The spec, to my reading, states that in-flight image resources should be safe when re-creating a swapchain - unless I'm getting this wrong?

**Comment by danilw** [comment]
@tgfrerer you did not mention but as I said - **this is only Nvidia Linux x11 issue**
And **only when this Nvidia used as Main GPU**.

**on Wayland and when Nvidia used as second GPU - everything fine**. 

From my user experience - "this crash" happen very often in "large applications" like games under DXVK, yes most of the games dont even allow you to resize window without restarting entire game but in some Unity games where you can resize window - it easy to get crash, so obviously as fix is just dont resize Vulkan window in x11 on Nvidia and set size on startup.

> Placing a vkQueueWaitIdle() before the (re-creating) call to vkCreateSwapchainKHR()

I also put this as fix last year, but then removed - it works inconsistent, still crash randomly, you just do less vkCreateSwapchainKHR calls so you need resize more times to do more calls... obviously random
**As I have seen** - even pausing rendering while user do resize, then waiting second after resize ended and only then do Swapchain recreation - it still crashes randomly so it completely RNG with `vkCreateSwapchainKHR` call.

**TLDR - this is only x11 Nvidia issue**, and I think can be considered as "wontfix" because x11 is almost deprecated at this point.

_Ofcourse if you have exact same issue and your case not happening because something else wrong in your Vulkan code_. I mean Vulkan is way too complex and impossible to debug realistically in something more complex than "single triangle app", in my simple single-triangle app it is clearly resize-crash come from "x11 Nvidia driver".

**Comment by tgfrerer** [comment]
@danilw thank you, that's interesting to hear! Yes, my environment is similar:  X11 and NVIDIA, with Nvidia being the main GPU; I'm curious to find out if it still happens on Wayland, I'll need to test this...

I've tested and encountered the issue with the same (`build/linux/app/bin/Release/x86_64/vulkan_samples --sample texture_loading`) application as in this original issue, and also with a ["spinning triangle" app](https://github.com/tgfrerer/island/tree/wip_swapchain_dynamic/apps/examples/hello_triangle) running with Island. 

It would be nice to have this resolved or hear something from the Nvidia folks who appear to be on this thread. I wonder if this issue should be raised with the Nvidia driver team directly (where?)... I'd be very happy to help with fixing this issue :)

**Comment by danilw** [comment]
@tgfrerer https://github.com/KhronosGroup/Vulkan-Samples/issues/250#issuecomment-868021052 this is Nvidia-dev response

**Comment by gpx1000** [comment]
I'm going to reopen this as it sounds like an NVIDIA bug and it's good to get visibility to them.  I recommend reporting it to their SDK team directly.

I apologize for not saying more before; but I have been unable to reproduce this.  I've tried both Wayland and X11 and haven't seen the crash.

**Comment by gpx1000** [comment]
I think we can close this after it gets visibility to NVIDIA.

**Comment by danilw** [comment]
> I have been unable to reproduce this.

**I think now I remember exactly how I fix it, and why**

## _this is not a Nvidia bug_
_this thread become very confusing_

This Vulkan-Samples handle "resize event" in Windows-way, **on Windows OS event `if (res == VK_ERROR_OUT_OF_DATE_KHR)` happening only "at end of render loop"** always and only at end.
So **this Vulkan-Samples source code does not handle `VK_ERROR_OUT_OF_DATE_KHR` event at beginning of render loop.**

When in Linux on x11 on Nvidia `VK_ERROR_OUT_OF_DATE_KHR` happens very **randomly at end or at beginning** of render loop, mostly at end this why crash is random.

Same with `VK_ERROR_SURFACE_LOST_KHR`, application should handle two this events at beginning and at the end of render loop, and not just at end as everyone do.

# I close this bug again, because this is not a Nvidia bug - this is "behavior" of Nvidia GPU driver on x11 that within Vulkan specs.

This is mistake of this Vulkan-Samples that spread across all Vulkan code, and it shows that you should not rely on any "source code of examples" and only read actual Vulkan-Specs because examples very often can "be broken/be bugged/promote not optimal way of handle events" (for example - this how Depth handled in this Vulkan-Samples is also source of errors/bugs in so many Vulkan apps that blindly copy/paste code).

___

@tgfrerer how **you can test that what I say is correct** - download my https://github.com/danilw/vulkan-shadertoy-launcher/releases or build it from source code, launch and resize - it will never crash. 

And I just tested this Vulkan-Samples as in my first post - `build/linux/app/bin/Release/x86_64/vulkan_samples sample texture_loading` does crash.

### Video recording of crash Vulkan-Samples just downloaded and build, crash at 34 sec on video https://youtu.be/62y3Oo6L0_M

**Comment by tgfrerer** [comment]
@danilw upon reflection, I take your point that the issue i've been seeing and the original issue are probably not be the same. 

* I would still like to be able to rule out that there might be an issue with the Nvidia Linux driver - perhaps here is the wrong place to put such a bug report, and I agree that things can get a bit confusing :) 

Can some of the Nvidia folks on here perhaps point me to a better place to file an issue? I'd be very happy to provide a test case etc... 

* In my code, I'm not depending on any resize callback at all to detect a window resize - instead, similar to @danilw's code, after I'm receiving `VK_ERROR_OUT_OF_DATE` on Present, the next frame will not acquire, but instead it will create a new swapchain, passing the old `VkSwapchain` as a parameter in order to re-use the Surface. Because my renderer can be triple-buffered, there might be an additional (CPU-)frame in-flight which is still referring to old images from the now retired swapchain. 
  That's why I keep the retired `VkSwapchain` alive for another triple-buffered round-trip before it is destroyed.
  From my understanding of the spec, this should be fine... However, if I don't issue a `VkQueueWaitIdle()` on the presenting queue - somewhere between immediately after I receive the first `VK_ERROR_OUT_OF_DATE` on present and just before I issue the call to `vkCreateSwapchainKHR` to re-create the swapchain - I get `VK_ERROR_DEVICE_LOST` more often than not on the next wait for a frame fence. 
  I understand that calling `Vk[Queue|Device]WaitIdle` when the swapchain is resized is perfectly fine from a user-experience point of view, but the fact that I *must* call it suggest to me that there might be something strange going on in the depths, and I'm curious to find out what it might be...

**Comment by damienleone** [comment]
@tgfrerer if you can send your repro app at vulkan-support@nvidia.com, I can take a look.

**Comment by danilw** [comment]
@tgfrerer 
> Because my renderer can be triple-buffered

I have example with multiple frames in flight for Linux https://github.com/danilw/nanovg-vulkan-glfw-integration-demo you have to build it, I dont have prebuild, you can set `const int MAX_FRAMES_IN_FLIGHT = 2;` in `example/example_vulkan_glfw_integration.cpp` source code, by default it 2.

It also should never crash. Test it if you want.

**Comment by jay3d** [comment]
@danilw Your app does not smooth resize on Windows. I did all the recommendations but still get the same crash.

**Comment by danilw** [comment]
> @danilw Your app does not smooth resize on Windows. 

I do not understand.

My application is - nanovg-vulkan-glfw-integration-demo I linked above?

And my application is "not smooth resize" for you? what this means? 
If you mean - "frame turning black and then repaint every time you resize" - this is how it suppose to be because I draw directly to "swapchain image" without caching framebuffer so swapchain re-render in real time on resize.

> I did all the recommendations but still get the same crash.

You have crash in my application nanovg-vulkan-glfw-integration-demo ?
Or you talk about your application still crashing?
If it your application - Vulkan is very complex with explosion of complexity.

So at this point you can start from making basic single triangle render in Vulkan that does not crash on resize, then start slowly adding new features from your exist application and step-by-step debug this way. 
It can be ginormous amount of time to debug it this way especially when crash random but I do not know other way.

**Or you can switch to using pre-build Vulkan engine, even Unity, and implement your logic there so it will not crash.** 
For same time of fixing your own engine you can just port it to engine that works, and start moving forward in production instead of infinite "debugging" overcomplicated mess that Vulkan created.
